src/wasm/parsing.cpp

```diff
@@ -105,6 +106,39 @@ void UniqueNameMapper::clear() {
 }
 
 void UniqueNameMapper::uniquify(Expression* curr) {
+  // First, scan the code to see if anything needs to be fixed up, since in the
+  // common case nothing needs fixing, and we can verify that very quickly.
+  struct Scanner
+    : public PostWalker<Scanner, UnifiedExpressionVisitor<Scanner>> {
+
+    // Whether things are ok. If not, we need to fix things up.
+    bool ok = true;
+
+    // It is rare to have many nested names in general, so track the seen names
+    // as we go in an efficient way.
+    SmallUnorderedSet<Name, 10> seen;
+
+    void visitExpression(Expression* curr) {
+      BranchUtils::operateOnScopeNameDefs(curr, [&](Name& name) {
```

Comment:
Does this traverse the entire expression tree rooted at `curr` or does it only visit `curr`? It would be unfortunate if a performance optimization accidentally introduced quadratic behavior!